stl/inc/xutility

```diff
@@ -6070,6 +6072,114 @@ private:
     _Udiff _Bmask; // 2^_Bits - 1
 };
 
+template <class _Diff, class _Urng>
+class _Rng_from_urng_v2 { // wrap a URNG as an RNG
+public:
+    using _Ty0 = make_unsigned_t<_Diff>;
+    using _Ty1 = _Invoke_result_t<_Urng&>;
+
+    using _Udiff                              = conditional_t<sizeof(_Ty1) < sizeof(_Ty0), _Ty0, _Ty1>;
+    static constexpr unsigned int _Udiff_bits = sizeof(_Udiff) * CHAR_BIT;
+    using _Uprod = conditional_t<_Udiff_bits <= 16, uint32_t, conditional_t<_Udiff_bits <= 32, uint64_t, _Unsigned128>>;
+
+    explicit _Rng_from_urng_v2(_Urng& _Func) : _Ref(_Func) {}
+
+    _Diff operator()(_Diff _Index) { // adapt _Urng closed range to [0, _Index)
+        // From Daniel Lemire, "Fast Random Integer Generation in an Interval", ACM Trans. Model. Comput. Simul. 29 (1),
+        // 2019.
+        _Udiff _Mask        = _Bmask;
+        unsigned int _Niter = 1;
+
+        if constexpr (_Bits < _Udiff_bits) {
+            while (_Mask < static_cast<_Udiff>(_Index - 1)) {
+                _Mask <<= _Bits;
+                _Mask |= _Bmask;
+                ++_Niter;
+            }
+        }
+
+        auto _Product = _Get_random_product(_Index, _Niter);
+        auto _Rem     = static_cast<_Udiff>(_Product) & _Mask;
+
+        if (_Rem < _Index) {
+            const auto _Threshold = (_Mask - (_Index - 1)) % _Index;

```

Comment:
Should this not be `(_Mask - _Index) % _Index`, or alternatively `_Mask % _Index`? In the original paper, I see

$t \leftarrow \left(2^L - s\right) \text{ mod } s$
and also
$\left\\{2^L \text{ mod } s = \left(2^L - s\right) \text{ mod } s\right\\}$

Reply:
$`_Mask` \equiv 2^L-1$, so we need to do `(_Mask+1) % _Index`, but that's not safe in general because it might overflow. We could `if constexpr` the case where it wouldn't overflow, if you think that would be worthwhile.

Reply:
I take it back, even if `_Bits < _Udiff_bits`, `_Mask` can still be all ones if `_Index` is large enough, so we can't know at compile time if the `inc` would be safe.

Reply:
Can this be `_Mask - _Index + 1` then? Additionally, `_Mask + 1 - _Index` would be fine, since `_Mask` is unsigned and thus overflow is defined.

Reply:
Ok, after reviewing the usual arithmetic conversions, I'm convinced that `_Index` will be converted to unsigned type if it isn't already. Therefore, an intermediate overflow is fine, as long as the final result is representable (which it is because $s < 2^L \leftrightarrow$ `_Index < _Mask+1`).